### PR TITLE
`Programming exercises`: Fix an issue with excess submissions with Gitlab

### LINKED ISSUE #5145: Some Programming Participations are not gradable
### Describe the bug

One of our courses encountered the problem that some participations (~10) for programming tasks are not gradable via artemis:

image

The participation in this example contains two submissions. 

image

In GitLab I see only one push event. Therefore, I assume that both commits are pushed via one push to the repository.

If you are looking on the dates, it seems that the first submission has a later date.
Nevertheless, the second submission is the second commit and should be gradable.
If I delete the first submission, it is gradable again (at least that was the case for the other cases)

I assume that could to be related to #4830 .

### To Reproduce

Unclear

### Expected behavior

The participation should be gradable.

### Screenshots

_No response_

### What browsers are you seeing the problem on?

Firefox

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [X] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [X] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).

### Motivation and Context

Closes #5145 
@dfuchss reported in #5145 that sometimes Artemis produces multiple submissions with different commits.
After investigation, I found out that this was the case if the student submits multiple commits within one push which happens rather rarely.

### Description

Gitlab provides in their [documentation ](https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html#push-events) an example of the payload of the push event webhook. The commits are sorted here in ascending order.

In our tests, however, we assumed a descending order. Because the previous implementation only took the first element on push-trigger, it always took the first commit, not the latest.

Gitlab provides the latest commit hash in a dedicated field. I took this hash and search for this specific commit. If it exists, which it usually does, we fill the response container with the details, otherwise, it only contains the hash itself.

@dfuchss confirmed that this fix resolves the issues he had.

Additionally, I added some test cases for this method and renamed the test request files in a more intuitive manner.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise
- Gitlab/Jenkins Setup

1. Create locally two commits and push both at the same time.
2. Make sure that only one submission got created with the commit hash of the latest commit

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| GitlabService.java | 50% | 83.66% |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabService.java [hunk 1]
```diff
@@ -329,14 +329,17 @@ public Boolean repositoryUrlIsValid(@Nullable VcsRepositoryUrl repositoryUrl) {
...
-        final var gitLabCommit = details.getCommits().get(0);
-        commit.setMessage(gitLabCommit.getMessage());
-        commit.setAuthorEmail(gitLabCommit.getAuthor().getEmail());
-        commit.setAuthorName(gitLabCommit.getAuthor().getName());
-        final var ref = details.getRef().split("/");
-        commit.setBranch(ref[ref.length - 1]);
-        commit.setCommitHash(gitLabCommit.getHash());
+        final var gitLabCommitHash = details.getNewHash();
+        commit.setCommitHash(gitLabCommitHash);
+        final var gitLabCommitOptional = details.getCommits().stream().filter(com -> gitLabCommitHash.equals(com.getHash())).findFirst();
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabService.java:334]: gitLabCommitOptional -> firstMatchingCommit

**Review by dfuchss — APPROVED** [review]
Fixes at least the bug regarding the ordering :)

**Review by rhubinger — APPROVED** [review]
Tested on TS2. Worked as described

**Review by JohannesStoehr — APPROVED** [review]
The code changes look good

**Review by yuling-sun — APPROVED** [review]
Tested on TS2. Works as described.

**Review by dfuchss — APPROVED** [review]
Code looks good.

**Pull request merged**